-#6909 Se creo un servicio para la generación de DOI

FacundoAdorno opened the pull request:
Este servicio es configurable. La configuracion actual evalua multiples condiciones, mediante un operador OR. Las condiciones son las siguientes, se pone en el estado de PARA GENERAR DOI a los items que cumplan con alguna de las siguientes:
* Si es una Tesis de posgrado o Tesis de maestria
* Si es un Libro de alguna editorial UNLP o similar (es decir, que empiece con 978-987-1985-, 978-987-4127-, 978-950-34-, 978-987-25535, 978-987-20132, 978-987-95358).
* Si es algun subtipo de conjunto de datos (conjunto de datos, placa espectrografica, placa fotografica).

FacundoAdorno committed:
Extracción de lógica de filtrado de DOI a Servicio independiente

Se extrajo la lógica para determinar si un item u objeto de DSpace es
elegible para generarle DOI a un servicio externo, de tal forma que
pueda utilzarse sencillamente desde cualquier otra clase.

FacundoAdorno committed:
Se agregó la base de la jerarquía de filtros configurable para el DOI Service Filter

Se agregaron las clases base necesarias para replicar la condición de
filtrado que estaba originalmente, es decir, que un item  tuviese un
subtype igual a "Tesis de grado" o "Tesis de Maestria" para que se le
genere el DOI.

FacundoAdorno committed:
-#6909 Se configuró el DOIFilterService

Para generarle DOIs a los Datasets, a los Libros de editoriales UNLP, ó
para tesis de grado y de maestria. Ademas se agregaron dos tipos de
filtros: un OR con multiple condiciones a evaluar y un AND con
multiples condiciones a evaluar.

arieljlira reviewed (commented):
@FacundoAdorno  HIce un comentario sobre el  MetadataComponentFilter y su uso. El resto lo veo bárbaro.

FacundoAdorno committed:
-#6909 Se corrigieron algunos filtros

Se implementó la negación para los filtros AND y OR, y además se eliminó
la propiedad "metadata" para moverlo a la subclase que lo utiliza.

FacundoAdorno commented on dspace/modules/additions/src/main/java/ar/edu/unlp/sedici/dspace/identifier/doi/filters/AbstractDOIFilter.java:
```diff
@@ -0,0 +1,33 @@
+    /**
+     * Contains the metadata element in the condition that determines partially if the filter applies.
+     */
+    protected MetadataComponentFilter metadata;
```
Realicé todas las modificaciones propuestas, excepto la de eliminar la clase MetadataComponentFilter, ya que me parece interesante tener esto desacoplado (por ahora), por si en un futuro es necesario crear otro filtro similar a HasSingleMetadataFilter.

FacundoAdorno commented:
El DoiConsumer solamente genera los DOIs en el momento del InstallItem. Si en un futuro, un ítem ya archivado que no tiene DOI es modificado y comienza a cumplir con las condiciones necesarias para generarsele un DOI, entonces el DOIConsumer no le genera un DOI. Es decir, que por ahora, el único momento en que un ítem puede generarsele DOI automáticamente es en el momento de la instalación del ítem.

santit96 commented on dspace/config/spring/api/identifier-service.xml:
```diff
@@ -42,17 +42,72 @@
+    <bean name="datasetFilter" class="ar.edu.unlp.sedici.dspace.identifier.doi.filters.HasSingleMetadataFilter">
+       <property name="metadata">
+           <bean class="ar.edu.unlp.sedici.dspace.identifier.doi.filters.MetadataComponentFilter">
+               <property name="metadataName" value="sedici.subtype"/>
```
Facu, aca no iría dc.type en vez de subtype?

santit96 commented on dspace/config/spring/api/identifier-service.xml:
```diff
@@ -42,17 +42,72 @@
+    <bean name="datasetFilter" class="ar.edu.unlp.sedici.dspace.identifier.doi.filters.HasSingleMetadataFilter">
+       <property name="metadata">
+           <bean class="ar.edu.unlp.sedici.dspace.identifier.doi.filters.MetadataComponentFilter">
+               <property name="metadataName" value="sedici.subtype"/>
```
Facu, acá no iría dc.type en vez del subtype? o se quieren hacer dois solo para el subtipo conjunto de datos y no para todo el tipo?

FacundoAdorno replied:
Si Santi, en realidad se quieren hacer solo para el subtipo "conjunto de datos" y no para el resto de los subtipos dentro de la misma tipología. Utilizar el "dc.type" es muy genérico, considerando que por ahí no todos los subtipos son de interés para generarle DOI. De todas formas, si en un futuro, quisieramos utilizar el dc.type se puede configurar fácilmente

santit96 reviewed (approved):
Facu, lo probé y anda todo joya. Lo unico que tengo una duda e el comentario ese que te puse. Los filtros funcionan como es esperado, probe distintos isbn y con distintos tipos de datos. Los DOIS que se generan a partir del submission quedan pendiente para el registro y despues tengo entendido que la cron definida en http://trac.prebi.unlp.edu.ar/issues/6049 se encargará de registrarlos.
